Add CODATA 2018 constants

pllim opened the pull request:
I revived #8167 but renamed 2017 to 2018, as discussed in https://github.com/astropy/astropy/pull/8167#issuecomment-444976586 . And I also revised the values based on CODATA values published in May 2018 at https://physics.nist.gov/cuu/Constants/index.html .

~~Are 2018 constants exactly the same as 2017? I am a bit confused here, please clarify.~~

I am gonna add a "what's new needed" label for now. Please remove if we don't want to advertise this in what's new.

xref #8155

mhvk commented:
Hmm, looking at this again, I recall that we're still waiting for the constants that are measured, and the statement that at least many of their measurement errors will change. Given that, should we wait for 3.2.1 after all?

bsipocz commented:
@mhvk - If the expectation is that those numbers come out on the 20th of May and we only need to plug them in, it's all good to go into 3.2 as the actual release is not expected to come out before that date (with the assumptions of RC will take at least 2 weeks after FF, and then ~month of RC testing).

pllim commented:
As discussed with @astrofrog offline, there is no reason why this cannot wait till 4.0. There is no reason to hold up 3.2 just for this.

stargaser commented:
I agree that this update is not appropriate for 3.2. I recommend to revisit this after CODATA2018 is out. For now the remilestone to 4.0 seems to me the right thing.

mhvk commented:
I'm still hopeful we can get this in as a "bug fix"... But milestone of 4.0 is fine for now, so that we don't get distracted.

bsipocz commented:
I'm with @mhvk here, and being fairly certain it doesn't even need to be a bugfix, v3.2 is 99% sure won't be out before the 20th of May.

pllim commented:
Update: @bsipocz mentioned that new values are out at https://physics.nist.gov/cuu/Constants/index.html . I'll update the values on this PR when I get the chance.

pllim commented:
@bsipocz , I updated the values and changed milestone back to 3.2. Feel free to re-milestone as needed.

pllim commented:
> what is next (perhaps we have it already?)

@mhvk , I think #8155 covers it.

mhvk commented:
Am not sure I understand what is next, i.e., why we cannot yet make this the default set of constants?

pllim commented:
@mhvk , I dunno, do we need to revisit https://en.wikipedia.org/wiki/2019_redefinition_of_the_SI_base_units and see if there are other changes in the `units` or `constants` code that will go with new defaults using CODATA 2018? Or do you think this *is* the extend of the impact as far as Astropy is concerned? Is 3.2 a proper release for new defaults or should we wait till 4.0?

stargaser commented:
From the experience of making CODATA2014 the new default, it is possible that tests in other packages will break and will need to be updated, if CODATA2018 is made the default.

I'm wondering if the kg being no longer a fundamental unit means something for decomposition of units...

stargaser commented:
To be clear: I personally was assuming we'd add CODATA2018 _and_ make it the default. I expect that to require some modification of existing tests.

pllim commented:
I am not sure if I am comfortable with sneaking in new defaults so close to 3.2 release, especially given 3.2rc1 has shipped. Do we need to discuss this offline? @astropy/coordinators , do you have any input?

mhvk commented:
I think I agree with @stargaser - we should have the latest and greatest in astropy 3.2! From a glance at the wikipedia article, I don't think it affects decomposition.

stargaser commented:
I missed the last comment from @mhvk while I was wrapping up my review. Yes, the latest and greatest should be in astropy 3.2. We must abide by the redefinition of SI! In another PR. Let's do it in two stages.

pllim commented on astropy/constants/codata2018.py:
```diff
@@ -0,0 +1,110 @@
+e = EMCODATA2018('e', 'Electron charge', 1.602176634e-19,
+                 'C', 0.0, system='si')
+
+eps0 = EMCODATA2018('eps0', 'Electric constant', 8.8541878128e-12,
```
I thought about it too, but am not sure how the human-readable name is used, and wonder if we need to deprecate or implement as alias. Regardless, I agree that renaming is a separate PR altogether.

pllim replied:
And for the record, I had to Google "electric constant" to find that it is a.k.a. "vacuum electric permittivity" and then enter the latter in the CODATA search box.

mhvk replied:
I'd vote for just renaming it. It is just the description of the constant, and I'm fairly sure ... permittivity is the right name.

pllim commented on astropy/constants/codata2018.py:
```diff
@@ -0,0 +1,110 @@
+
+# Formula taken from NIST wall chart.
+b_wien = CODATA2018('b_wien', 'Wien wavelength displacement law constant',
+                    h.value * c.value / (k_B.value * 4.965114231), 'm K',
```
Done!

mhvk replied:
Could you add the link to the wiki page? I had been wondering about this too, but was too lazy to comment (thanks for doing a much more careful job, @stargaser!)

pllim commented:
> Let's do it in two stages.

Just want to clarify -- Do you mean making it default as a separate PR? Or did you mean something else?

mhvk commented on astropy/constants/codata2018.py:
```diff
@@ -0,0 +1,110 @@
+atm = CODATA2018('atm', "Standard atmosphere", 101325,
+                 'Pa', 0.0, system='si')
+
+mu0 = CODATA2018('mu0', "Magnetic constant", 1.25663706212e-6,
```
Yes, let's. "constant" is just vague.

mhvk commented:
@pllim - could you still do the name changes to permittivity and the link to the wikipedia page for the `b_wien` constant?  Otherwise, it looks all OK (and I am happy with @stargaser's suggestion to stick with your plan of getting this in first, and then worry about making it the default).

stargaser commented:
I concur with @mhvk. I meant, let's make a separate PR for making CODATA2018 the default. I agree the name changes to permittivity should be in this PR, and don't require deprecation. The description is just helpful text for users, and really, these are new constants that we are adding.

pllim commented:
@stargaser and @mhvk , thank you very much for the reviews! I think I have addressed all the comments. And yes, I also concur that making it default can be a separate PR.

mhvk commented:
Thanks, @pllim! Merging...
